File: 단계별로 풀어보기/백트래킹/15650/N과 M (2).py
import sys

def dfs(start, cnt):
    
    if cnt == M:
        for i in range(M):
            sys.stdout.write('%d ' %arr[i])
        sys.stdout.write('\n')
        
    for i in range(start, N):
        if not visited[i]:
            visited[i] = True
            arr[cnt] = i+1
            start = i
            dfs(start, cnt+1)
            visited[i] = False

N, M = map(int, sys.stdin.readline().split())
arr = [0 for i in range(8)]
visited = [False for i in range(8)]
dfs(0, 0)

File: 단계별로 풀어보기/백트래킹/15649/N과 M (2).cpp
#include <iostream>
#define MAX 8
using namespace std;

int N, M;
int arr[MAX] = {0,};
bool visited[MAX] = {0,};

int dfs(int start, int cnt){
    if(cnt == M){
        for(int i = 0; i < M; i++){
            cout << arr[i]+1 << ' ';
        }
        
        cout << '\n';
    }
    
    for(int i = start; i < N; i++){
        if(!visited[i]){
            visited[i] = true;
            arr[cnt]=i;
            start = i;
            dfs(start, cnt+1);
            visited[i] = false;
        }
    }
    
    return 0;
}



int main(){
    cin >> N >> M;
    dfs(0,0);
    return 0;
}
